[Windows] Crashes with no stacktrace

The DFS and heat benchmarks are crashing with no stacktrace on Windows: https://github.com/status-im/nim-taskpools/runs/2962435403#step:13:89

```
========================================================================================
Running [ c  ] benchmarks/heat/taskpool_heat.nim
========================================================================================
[NimScript] exec: nim c  --verbosity:0 --hints:off --warnings:off --threads:on -d:release --stacktrace:on --linetrace:on --outdir:build -r benchmarks/heat/taskpool_heat.nim
Error: execution of an external program failed: 'D:\a\nim-taskpools\nim-taskpools\nim-taskpools\build\taskpool_heat.exe '
stack trace: (most recent call last)
C:\Users\RUNNER~1\AppData\Local\Temp\nimblecache-0\nimscriptapi_2697630897.nim(187, 16)
D:\a\nim-taskpools\nim-taskpools\nim-taskpools\taskpools.nimble(37, 8) testTask
D:\a\nim-taskpools\nim-taskpools\nim-taskpools\taskpools.nimble(26, 8) test
D:\a\nim-taskpools\nim-taskpools\nim\lib\system\nimscript.nim(260, 7) exec
D:\a\nim-taskpools\nim-taskpools\nim\lib\system\nimscript.nim(260, 7) Error: unhandled exception: FAILED: nim c  --verbosity:0 --hints:off --warnings:off --threads:on -d:release --stacktrace:on --linetrace:on --outdir:build -r benchmarks/heat/taskpool_heat.nim [OSError]
       Tip: 1 messages have been suppressed, use --verbose to show them.
     Error: Exception raised during nimble script execution
Error: Process completed with exit code 1.
```

Looking locally with gdb we get this

```
Thread 36 received signal SIGTRAP, Trace/breakpoint trap.
[Switching to Thread 5752.0x24d8]
0x00007ffde2608c07 in ntdll!RtlpNtMakeTemporaryKey () from C:\windows\SYSTEM32\ntdll.dll
(gdb) bt
#0  0x00007ffde2608c07 in ntdll!RtlpNtMakeTemporaryKey () from C:\windows\SYSTEM32\ntdll.dll
#1  0x00007ffde25f5c85 in ntdll!RtlZeroHeap () from C:\windows\SYSTEM32\ntdll.dll
#2  0x00007ffde25c49db in ntdll!memset () from C:\windows\SYSTEM32\ntdll.dll
#3  0x00007ffde2607a94 in ntdll!RtlpNtMakeTemporaryKey () from C:\windows\SYSTEM32\ntdll.dll
#4  0x00007ffde25aba02 in ntdll!memset () from C:\windows\SYSTEM32\ntdll.dll
#5  0x00007ffde250eb2e in ntdll!RtlFreeHeap () from C:\windows\SYSTEM32\ntdll.dll
#6  0x00007ffde08d9d0c in msvcrt!free () from C:\windows\System32\msvcrt.dll
#7  0x000000000042b8d2 in cleanup__SHKET3vaU9a9cCo73MoYsGXgflowvars (fv=...) at [...]/nim-taskpools/taskpools/flowvars.nim:41
#8  sync__wTuc87TnWJbDfXB9bbnSyFgflowvars (fv=...) at [...]/nim-taskpools/taskpools/flowvars.nim:72
#9  heat__RQB9b6xKuruBDMMs9apUS0MA (m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>,
    il@entry=3582, iu=<optimized out>, iu@entry=0) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:144
#10 0x000000000042b760 in heat__RQB9b6xKuruBDMMs9apUS0MA (
    m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>, il@entry=3580,
    iu=<optimized out>, iu@entry=0) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:143
#11 0x000000000042b760 in heat__RQB9b6xKuruBDMMs9apUS0MA (
    m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>, il@entry=3576,
    iu=<optimized out>, iu@entry=0) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:143
#12 0x000000000042b760 in heat__RQB9b6xKuruBDMMs9apUS0MA (
    m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>, il@entry=3568,
    iu=<optimized out>, iu@entry=0) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:143
#13 0x000000000042b760 in heat__RQB9b6xKuruBDMMs9apUS0MA (
    m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>, il@entry=3552,
    iu=<optimized out>, iu@entry=0) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:143
#14 0x000000000042b760 in heat__RQB9b6xKuruBDMMs9apUS0MA (
    m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>, il@entry=3520,
    iu=<optimized out>, iu@entry=0) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:143
#15 0x000000000042b760 in heat__RQB9b6xKuruBDMMs9apUS0MA (
    m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>, il@entry=3456,
    iu=<optimized out>, iu@entry=0) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:143
#16 0x000000000042b760 in heat__RQB9b6xKuruBDMMs9apUS0MA (
    m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>, il@entry=3328,
    iu=<optimized out>, iu@entry=176570576) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:143
#17 0x000000000042b760 in heat__RQB9b6xKuruBDMMs9apUS0MA (
    m=<error reading variable: Cannot access memory at address 0x2b1578c7c5620000>, il=<optimized out>, il@entry=3072,
    iu=<optimized out>, iu@entry=3584) at [...]/nim-taskpools/benchmarks/heat/taskpool_heat.nim:143
#18 0x000000000042db4e in taskpool_heat__U5YlCX8Yrr5EIVO1DlnNSQ (m=..., il=3072, iu=3584, fut=...)
    at [...]/nim-taskpools/taskpools/taskpools.nim:497
#19 0x000000000042e376 in taskpool_heat__5BGkgO9b76lIYZk7e9ccCl9cg (argsX60gensym17100250_=0xb251a50)
    at [...]/nim-taskpools/taskpools/shims_pre_1_6/tasks.nim:130
#20 0x0000000000425a08 in invoke__h7ET9aV09bIhD9aJ85VPycUJQtasks (task=...)
```

--- a/taskpools.nimble
+++ b/taskpools.nimble
@@ -34,15 +34,13 @@ task test, "Run Taskpools tests":
   test "", "examples/e01_simple_tasks.nim"
 
   # Benchmarks
+  test "", "benchmarks/dfs/taskpool_dfs.nim"
   test "", "benchmarks/heat/taskpool_heat.nim"
   test "", "benchmarks/nqueens/taskpool_nqueens.nim"
 
   when not defined(windows):
     test "", "benchmarks/single_task_producer/taskpool_spc.nim"
     test "", "benchmarks/bouncing_producer_consumer/taskpool_bpc.nim"
-
-    # TODO - unsure why this fails on windows
-    test "", "benchmarks/dfs/taskpool_dfs.nim"
 
   # TODO - generics in macro issue
   # test "", "benchmarks/matmul_cache_oblivious/taskpool_matmul_co.nim"

--- a/taskpools/flowvars.nim
+++ b/taskpools/flowvars.nim
@@ -6,10 +6,11 @@
 # at your option. This file may not be copied, modified, or distributed except according to those terms.
 
 import
-  ./channels_spsc_single,
   system/ansi_c,
+  std/os,
   ./instrumentation/contracts,
-  std/os
+  ./channels_spsc_single,
+  ./primitives/allocs
 
 {.push gcsafe.}
 
@@ -32,13 +33,13 @@ type
 
 proc newFlowVar*(T: typedesc): Flowvar[T] {.inline.} =
   let size = 2 + sizeof(T) # full flag + item size + buffer
-  result.chan = cast[ptr ChannelSPSCSingle](c_calloc(1, csize_t size))
+  result.chan = wv_allocAligned(ChannelSPSCSingle, size, alignment = 64)
   result.chan[].initialize(sizeof(T))
 
 proc cleanup(fv: Flowvar) {.inline.} =
   # TODO: Nim v1.4+ can use "sink Flowvar"
   if not fv.chan.isNil:
-    c_free(fv.chan)
+    wv_freeAligned(fv.chan)
 
 func isSpawned*(fv: Flowvar): bool {.inline.} =
   ## Returns true if a flowvar is spawned

--- a/taskpools/primitives/allocs.nim
+++ b/taskpools/primitives/allocs.nim
@@ -1,0 +1,152 @@
+# Weave
+# Copyright (c) 2019 Mamy André-Ratsimbazafy
+# Licensed and distributed under either of
+#   * MIT license (license terms in the root directory or at http://opensource.org/licenses/MIT).
+#   * Apache v2 license (license terms in the root directory or at http://www.apache.org/licenses/LICENSE-2.0).
+# at your option. This file may not be copied, modified, or distributed except according to those terms.
+
+import
+  system/ansi_c
+
+# Helpers
+# ----------------------------------------------------------------------------------
+
+proc isPowerOfTwo*(n: int): bool {.inline.} =
+  (n and (n - 1)) == 0
+
+# TODO: cannot dispatch at compile-time due to https://github.com/nim-lang/Nim/issues/12726
+# but all our use-case are for power of 2
+
+func roundNextMultipleOf*(x: Natural, n: Natural): int {.inline.} =
+  assert n.isPowerOfTwo()
+  result = (x + n - 1) and not(n - 1)
+
+# func roundNextMultipleOf*(x: Natural, n: static Natural): int {.inline.} =
+#   ## Round the input to the next multiple of "n"
+#   when n.isPowerOfTwo():
+#     # n is a power of 2. (If compiler cannot prove that x>0 it does not make the optim)
+#     result = (x + n - 1) and not(n - 1)
+#   else:
+#     result = ((x + n - 1) div n) * n
+
+# Memory
+# ----------------------------------------------------------------------------------
+
+# Nim allocShared, createShared, deallocShared
+# take a global lock that is absolutely killing performance
+# and shows up either:
+# - native_queued_spin_lock_slowpath
+# - __pthread_mutex_lock and __pthread_mutex_unlock_usercnt
+#
+# We use system malloc by default, the flag -d:useMalloc is not enough
+
+template deref*(T: typedesc): typedesc =
+  ## Return the base object type behind a ptr type
+  typeof(default(T)[])
+
+proc wv_alloc*(T: typedesc): ptr T {.inline.}=
+  ## Default allocator for the Picasso library
+  ## This allocates memory to hold the type T
+  ## and returns a pointer to it
+  ##
+  ## Can use Nim allocator to measure the overhead of its lock
+  ## Memory is not zeroed
+  when defined(WV_useNimAlloc):
+    createSharedU(T)
+  else:
+    cast[ptr T](c_malloc(csize_t sizeof(T)))
+
+proc wv_allocPtr*(T: typedesc[ptr], zero: static bool = false): T {.inline.}=
+  ## Default allocator for the Picasso library
+  ## This allocates memory to hold the
+  ## underlying type of the pointer type T.
+  ## i.e. if T is ptr int, this allocates an int
+  ##
+  ## Can use Nim allocator to measure the overhead of its lock
+  ## Memory is not zeroed
+  result = wv_alloc(deref(T))
+  when zero:
+    zeroMem(result, sizeof(deref(T)))
+
+proc wv_alloc*(T: typedesc, len: SomeInteger): ptr UncheckedArray[T] {.inline.} =
+  ## Default allocator for the Picasso library.
+  ## This allocates a contiguous chunk of memory
+  ## to hold ``len`` elements of type T
+  ## and returns a pointer to it.
+  ##
+  ## Can use Nim allocator to measure the overhead of its lock
+  ## Memory is not zeroed
+  when defined(WV_useNimAlloc):
+    cast[type result](createSharedU(T, len))
+  else:
+    cast[type result](c_malloc(csize_t len*sizeof(T)))
+
+proc wv_free*[T: ptr](p: T) {.inline.} =
+  when defined(WV_useNimAlloc):
+    freeShared(p)
+  else:
+    c_free(p)
+
+when defined(windows):
+  proc alloca(size: int): pointer {.header: "<malloc.h>".}
+else:
+  proc alloca(size: int): pointer {.header: "<alloca.h>".}
+
+template alloca*(T: typedesc): ptr T =
+  cast[ptr T](alloca(sizeof(T)))
+
+template alloca*(T: typedesc, len: Natural): ptr UncheckedArray[T] =
+  cast[ptr UncheckedArray[T]](alloca(sizeof(T) * len))
+
+when defined(windows):
+  proc aligned_alloc_windows(size, alignment: csize_t): pointer {.sideeffect,importc:"_aligned_malloc", header:"<malloc.h>".}
+    # Beware of the arg order!
+  proc wv_freeAligned*[T](p: ptr T){.sideeffect,importc:"_aligned_free", header:"<malloc.h>".}
+elif defined(osx):
+  proc posix_memalign(mem: var pointer, alignment, size: csize_t){.sideeffect,importc, header:"<stdlib.h>".}
+  proc aligned_alloc(alignment, size: csize_t): pointer {.inline.} =
+    posix_memalign(result, alignment, size)
+  proc wv_freeAligned*[T](p: ptr T){.inline.} =
+    c_free(p)
+else:
+  proc aligned_alloc(alignment, size: csize_t): pointer {.sideeffect,importc, header:"<stdlib.h>".}
+  proc wv_freeAligned*[T](p: ptr T){.inline.} =
+    c_free(p)
+
+proc wv_allocAligned*(T: typedesc, alignment: static Natural): ptr T {.inline.} =
+  ## aligned_alloc requires allocating in multiple of the alignment.
+  static:
+    assert alignment.isPowerOfTwo()
+  let # TODO - cannot use a const due to https://github.com/nim-lang/Nim/issues/12726
+    size = sizeof(T)
+    requiredMem = size.roundNextMultipleOf(alignment)
+
+  when defined(windows):
+    cast[ptr T](aligned_alloc_windows(csize_t requiredMem, csize_t alignment))
+  else:
+    cast[ptr T](aligned_alloc(csize_t alignment, csize_t requiredMem))
+
+proc wv_allocAligned*(T: typedesc, size: int, alignment: static Natural): ptr T {.inline.} =
+  ## aligned_alloc requires allocating in multiple of the alignment.
+  static:
+    assert alignment.isPowerOfTwo()
+  let
+    requiredMem = size.roundNextMultipleOf(alignment)
+
+  when defined(windows):
+    cast[ptr T](aligned_alloc_windows(csize_t requiredMem, csize_t alignment))
+  else:
+    cast[ptr T](aligned_alloc(csize_t alignment, csize_t requiredMem))
+
+proc wv_allocArrayAligned*(T: typedesc, len: int, alignment: static Natural): ptr UncheckedArray[T] {.inline.} =
+  ## aligned_alloc requires allocating in multiple of the alignment.
+  static:
+    assert alignment.isPowerOfTwo()
+  let
+    size = sizeof(T) * len
+    requiredMem = size.roundNextMultipleOf(alignment)
+
+  when defined(windows):
+    cast[ptr UncheckedArray[T]](aligned_alloc_windows(csize_t requiredMem, csize_t alignment))
+  else:
+    cast[ptr UncheckedArray[T]](aligned_alloc(csize_t alignment, csize_t requiredMem))

--- a/taskpools/taskpools.nim
+++ b/taskpools/taskpools.nim
@@ -43,7 +43,7 @@ import
   ./channels_spsc_single,
   ./chase_lev_deques,
   ./event_notifiers,
-  ./primitives/barriers,
+  ./primitives/[barriers, allocs],
   ./instrumentation/[contracts, loggers],
   ./sparsesets,
   ./flowvars,
@@ -98,6 +98,7 @@ type
   Taskpool* = ptr object
     barrier: SyncBarrier
       ## Barrier for initialization and teardown
+    # --- Align: 64
     eventNotifier: EventNotifier
       ## Puts thread to sleep
 
@@ -348,14 +349,16 @@ proc new*(T: type Taskpool, numThreads = countProcessors()): T {.raises: [Except
   ## Initialize a threadpool that manages `numThreads` threads.
   ## Default to the number of logical processors available.
 
-  var tp = cast[T](c_calloc(1, csize_t sizeof(default(Taskpool)[])))
+  type TpObj = typeof(default(Taskpool)[])
+  # Event notifier requires an extra 64 bytes for alignment
+  var tp = wv_allocAligned(TpObj, sizeof(TpObj) + 64, 64)
 
   tp.barrier.init(numThreads.int32)
   tp.eventNotifier.initialize()
   tp.numThreads = numThreads
-  tp.workerDeques = cast[ptr UncheckedArray[ChaseLevDeque[TaskNode]]](c_calloc(csize_t numThreads, csize_t sizeof ChaseLevDeque[TaskNode]))
-  tp.workers = cast[ptr UncheckedArray[Thread[(Taskpool, WorkerID)]]](c_calloc(csize_t numThreads, csize_t sizeof Thread[(Taskpool, WorkerID)]))
-  tp.workerSignals = cast[ptr UncheckedArray[Signal]](c_calloc(csize_t numThreads, csize_t sizeof Signal))
+  tp.workerDeques = wv_allocArrayAligned(ChaseLevDeque[TaskNode], numThreads, alignment = 64)
+  tp.workers = wv_allocArrayAligned(Thread[(Taskpool, WorkerID)], numThreads, alignment = 64)
+  tp.workerSignals = wv_allocArrayAligned(Signal, numThreads, alignment = 64)
 
   # Setup master thread
   workerContext.id = 0
@@ -397,13 +400,13 @@ proc cleanup(tp: var TaskPool) {.raises: [OSError].} =
   for i in 1 ..< tp.numThreads:
     joinThread(tp.workers[i])
 
-  tp.workerSignals.c_free()
-  tp.workers.c_free()
-  tp.workerDeques.c_free()
+  tp.workerSignals.wv_freeAligned()
+  tp.workers.wv_freeAligned()
+  tp.workerDeques.wv_freeAligned()
   `=destroy`(tp.eventNotifier)
   tp.barrier.delete()
 
-  tp.c_free()
+  tp.wv_freeAligned()
 
 proc shutdown*(tp: var TaskPool) {.raises:[Exception].} =
   ## Wait until all tasks are processed and then shutdown the taskpool

